src/envoy/http/authn/filter_context.cc

```diff
@@ -74,24 +74,47 @@ void FilterContext::setPrincipal(const iaapi::PrincipalBinding& binding) {
 
 bool FilterContext::getJwtPayload(const std::string& issuer,
                                   std::string* payload) const {
-  const auto filter_it =
-      dynamic_metadata_.filter_metadata().find(Utils::IstioFilterName::kJwt);
+  // Try getting the Jwt payload from Envoy jwt_authn filter.
+  // If not found, try getting the Jwt payload from Istio jwt-auth filter.
+  bool isUsingEnvoyJwtFilter = false;
+  auto filter_it =
+      dynamic_metadata_.filter_metadata().find(Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
   if (filter_it == dynamic_metadata_.filter_metadata().end()) {
     ENVOY_LOG(debug, "No dynamic_metadata found for filter {}",
-              Utils::IstioFilterName::kJwt);
-    return false;
+              Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
+    ENVOY_LOG(debug, "Now try fetching dynamic_metadata for filter {}",
+             Utils::IstioFilterName::kJwt);
+    filter_it = dynamic_metadata_.filter_metadata().find(Utils::IstioFilterName::kJwt);
+    if (filter_it == dynamic_metadata_.filter_metadata().end()) {
+      ENVOY_LOG(debug, "No dynamic_metadata found for filter {}",
+                Utils::IstioFilterName::kJwt);
+      return false;
+    }
+  } else {
+    isUsingEnvoyJwtFilter = true;
```

Comment:
It's a bit hard to read. I suggest to have a separate getJwtPayload that deal with only upstream JWT filter, and one with Istio Jwt (i.e this one, but rename). This function then can simply be

```
bool FilterContext::getJwtPayload(const std::string& issuer,
                                  std::string* payload) const {	                                  std::string* payload) const {
  return getJwtPayloadFromUpstreamJwt(issuer, payload) || getJwtPayloadFromIstioJwt(...);
}
```

Reply:
Sounds good. Thanks!

Reply:
Done. PTAL.